apps/scratch/src/utilities/scratch-block.ts

```diff
@@ -0,0 +1,80 @@
+import { WorkspaceChangeEvent } from "../types/scratch-workspace";
+import type { Block } from "scratch-blocks";
+
+const minStackSize = 2;
+
+export type DeletedBlockInfo = Array<{ id: string; type: string }>;
+export interface DeletedBlockRecord {
+  id: string;
+  type: string;
+  blockIdArray: DeletedBlockInfo;
+}
+
+function getStackSize(topBlock: Block): number {
+  let size = 0;
+  let current: Block | null = topBlock;
+
+  while (current) {
+    size++;
+    current = current.getNextBlock?.() ?? null;
+  }
+  return size;
+}
+
+export function isBlockPartOfLargeStack(block: Block): boolean {
+  if (!block) {
+    return false;
+  }
+
+  let current: Block | null = block;
+
+  while (current?.getParent?.()) {
+    current = current.getParent?.() ?? null;
+  }
+  if (!current) {
+    return false;
+  }
+
+  const size = getStackSize(current);
+  return size >= minStackSize;
+}
+
+export const wasPartOfLargeStack = (event: WorkspaceChangeEvent): boolean => {
+  return event.oldParentId != null;
+};
+
+export const mapXmlBlockToBlock = (
+  xmlElement: Element,
+): DeletedBlockRecord | null => {
+  // Extract the block ID and type from the XML element
+  const blockId = xmlElement.getAttribute("id");
+  const blockType = xmlElement.getAttribute("type");
+
+  if (!blockId || !blockType) {
+    return null;
```

Comment:
In general you return `null` in a lot of places and when reviewing it is unclear whether this are codepaths that are ever supposed to be taken. If so, then keep it but otherwise throw exceptions instead.